### DISCUSSION TITLE
Social part of dependency management: visualization of technical debt

### DISCUSSION THREAD

**Discussion by MShekow** [renovatebot/renovate#26024] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

This is _not_ a _technical_ question, so I apologize if this is not the right place to ask.

We are using Renovate Bot in many projects. IMO Renovate is the best dependency management software, compared to competing products (dependabot, etc.). 

When introducing Renovate, I observe a common pattern, unfortunately:

1. Someone (usually me) adds Renovate to a software project whose team has been focusing on building _features_ for the past years, mostly ignoring updating dependencies
2. Renovate finds 100+ outdated dependencies
3. I see disbelieve in everyone's face, paralysis, shock, uncertainty what to do with this result

I'm interested how other teams handle this kind of situation. How do you handle questions like:

- How do you handle a large tech-debt-backlog (meaning: _many_ dependency-related PRs)? How do you prioritize? How do you plan for _regular_ maintenance (e.g. as part of a Scrum sprint)?
- How do you convince the _non-technical_ stakeholders (who _pay_ for everything) why regular dependency updates are important?
- How do you keep track of the (open) Renovate PRs over time? We are thinking about writing some code that generates graphs of the open PRs over time (possibly separately showing major vs. minor/patch vs. security updates), as this would show whether the technical-debt-situation deteriorates or improves. However, I did not find any existing tooling for generating such graphs.

What I/we usually do:

- Calm the team down. The team needs to understand that we have now just established _transparency_, with _no_ immediate need to act _now_. Teams feel much calmer if they understand that, at least initially, these PRs are just for planning and transparency. In some cases, it can even make sense to keep Renovate's onboarding PR open, as it already provides enough information (and having Renovate generate all individual PRs has no advantage yet).
  - If you don't calm the team down, they might just uninstall Renovate again (putting the lid back on Pandora's box, so to speak :D)
- If the team decides to merge the onboarding PR, I teach the team how to create a browser bookmark that filters out the PRs created by Renovate, based on a _label_ (as otherwise the team won't find _their own_ PRs anymore in the "sea of PRs", which would cause strong irritation for the team, making them "hate" Renovate right away)
- As for the advantages of _regular_ updates, I quote from a [Renovate introduction blog post](https://www.augmentedmind.de/2023/07/30/renovate-bot-introduction/) of mine:
  - The increment of each individual upgrade is small (e.g. just a minor version update), which means that the update bears _less risk_ of breaking your application.
  - It is _easier to plan_ the work, because integrating small increments is less work.
  - The chances of being affected by _security vulnerabilities_ is smaller, if you update to the latest version of a dependency, shortly after it has been released.
  - It pushes your team towards _building better automated tests_, because then you can configure Renovate to fully automatically merge successfully-tested PRs, reducing the maintenance work even further.

I'm curious about your opinions.

**Comment by HonkingGoose** [comment]
This is a very good topic! I already wrote our best practices into the [Renovate docs, upgrade best practices](https://docs.renovatebot.com/upgrade-best-practices/) page. I recommend you read that page first, to get a overview of my thoughts.

You're asking a lot of questions, so I'll try to get back to you in-depth later.

**Comment by HonkingGoose** [comment, marked as answer]
## Answering your questions

First, I don't have experience working with a big team, and usually my team was small, and I dealt with the updating stuff. So I usually got what I wanted anyways... 😄 So remember that when you read my answers. 😉 

> * How do you handle a large tech-debt-backlog (meaning: _many_ dependency-related PRs)?

I'd spend a Scrum sprint cycle on basic maintenance:

- Updating "easy" dependencies
- Preparing to update difficult dependencies
- Refactoring code
- Updating or writing tests

> How do you prioritize?

Do the known-bad stuff first, so vulnerability stuff first.

> How do you plan for _regular_ maintenance (e.g. as part of a Scrum sprint)?

_Improve_ the [Definition of Done as per the Scrum guide](https://scrumguides.org/scrum-guide.html#increment), so that "Done" includes things like:

- No known vulnerabilities
- Up-to-date stack
- Clean code

> * How do you convince the _non-technical_ stakeholders (who _pay_ for everything) why regular dependency updates are important?

In the EU there are laws that allow oversight commities to fine companies for data leaks. Those fines can get quity hefty! So there's a financial reason. Then there's goodwill/reputation loss due to data leaks.

And it's clearly the right thing to care about the security of your software, to keep your customers safe.

> * How do you keep track of the (open) Renovate PRs over time? We are thinking about writing some code that generates graphs of the open PRs over time (possibly separately showing major vs. minor/patch vs. security updates), as this would show whether the technical-debt-situation deteriorates or improves. However, I did not find any existing tooling for generating such graphs.

I think these things can help and inspire you:

- https://github.com/renovatebot/renovate/discussions/23105
- https://github.com/renovatebot/renovate/issues/23961
- [Renovate docs, Swissquote user story that has a cool custom Renovate dashboard](https://docs.renovatebot.com/user-stories/swissquote/)

**Reply by MShekow** [reply]
Thanks for all the input.

Some of my takeaways (focusing on the human aspects):

- Convince the stakeholders on the business side to allow the development team to allocate a certain time budget for updating dependencies (e.g. x hours per sprint). Convincing arguments should include: "avoid horribly high costs due to a breach caused by vulnerabilities" but also "keep a (roughly) _constant development velocity_ by avoiding _unplanned_ work". For the latter, it would probably be good to present a nice collection of horror stories where unplanned work threw the team off.
- Establish a general rule (e.g. as part of the Definition of Done) that _vulnerability-related_ PRs must always be dealt with right away, even if the "dependency update time budget" is already depleted, meaning that you need to "drop everything" to address vulnerabilities (at least those with a certain minimum CVSS score)
- For teams who just introduced Renovate into a project with many outdated dependencies: establish guidelines for prioritization. This might include hints like "update easy (minor/patch) dependencies first, where you estimate a low risk of breakage", but could also mean to prioritize creating more automated tests _first_: while this would _initially_ slow down the speed of "merged dependency PRs per sprint", you would save a lot of time in the long run, and it could go as far as "automerge=true" for almost every kind of dependency. So in some sense, I disagree with the statement _"You're done with the technical side. Now comes the harder part, fixing the human side"_ made [here](https://docs.renovatebot.com/upgrade-best-practices/#improve-the-human-side) in the _upgrade best practices guide_, as I would attempt to look at the _human_ side first.

It seems to me that having dashboards regarding Renovate's use is definitely worth it (as they may help when making a point for introducing fixing-dependency-budgets). The dashboards mention in your links revolve around _technical_ aspects (e.g. Prometheus metrics regarding the _duration_ of certain operations). So it seems there is still some work to do (including, but not limited to, https://github.com/renovatebot/renovate/issues/23961) to build more "managerial" dashboards, showing data such as:
- How many of our repos already use Renovate (i.e., onboarding completed), or have it disabled completely (no open onboarding PR, no renovate.json file in "main" either)
- For those repos still in the onboarding phase: how many PRs would Renovate create
- PR statistics for already-onboarded repos, e.g. number of open PRs over time, average time until a PR is closed (sub-divided by version-upgrade level: major/minor/patch/vulnerability)

Also, I guess that having automated _preview environments_ (for web-based apps) will help the team a lot, during the "manual testing" phase. Otherwise, it's time-intensive to always having to check out some Renovate-created branch, building the app locally, then starting it. However, because preview environments cost money/resources, one will need to build some kind of mechanism to limit the preview environments to only specific Renovate PRs (in case you have many).

**Reply by MShekow** [reply]
By the way, I built a prototype for visualizing Renovate PRs, feel free to check it out: https://github.com/MShekow/renovate-pr-visualization
